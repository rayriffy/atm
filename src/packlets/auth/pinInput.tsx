import {
  HStack,
  PinInput as ChakraPinInput,
  PinInputField,
  Alert,
  AlertIcon,
} from '@chakra-ui/react'
import { Fragment } from 'react'

import { usePinInput } from './usePinInput'

export const PinInput = () => {
  const { pin, setPin, loading, error } = usePinInput()

  return (
    <Fragment>
      <HStack py={12} justify="center">
        <ChakraPinInput
          size="lg"
          mask
          variant="flushed"
          value={pin}
          onChange={setPin}
          isDisabled={loading}
          isInvalid={error}
        >
          <PinInputField />
          <PinInputField />
          <PinInputField />
          <PinInputField />
        </ChakraPinInput>
      </HStack>
      {error && (
        <Alert status="error" variant="solid" rounded="md">
          <AlertIcon />
          Pin is incorrect
        </Alert>
      )}
    </Fragment>
  )
}
