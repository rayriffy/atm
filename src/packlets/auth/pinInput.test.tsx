import { render, act, fireEvent } from '@testing-library/react'
import { vi, describe, it, expect, beforeEach } from 'vitest'

import { PinInput } from './pinInput'

const setPinSpy = vi.hoisted(() => vi.fn())
const usePinInputSpy = vi.hoisted(() =>
  vi.fn(() => ({
    pin: '',
    setPin: setPinSpy,
    loading: false,
    error: false,
  }))
)
vi.mock('./usePinInput', () => ({
  usePinInput: usePinInputSpy,
}))

describe('<PinInput />', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('should render and match snapshot', () => {
    const { container } = render(<PinInput />)

    expect(container).toMatchSnapshot()
  })

  it('should call setPin when type', () => {
    const { container } = render(<PinInput />)

    const input = container.querySelector('input')

    act(() => {
      fireEvent.change(input!, { target: { value: '1' } })
    })

    expect(setPinSpy).toBeCalledTimes(1)
    expect(setPinSpy).toHaveBeenCalledWith('1')
  })

  it('should disable input when loading', () => {
    usePinInputSpy.mockReturnValueOnce({
      pin: '',
      setPin: setPinSpy,
      loading: true,
      error: false,
    })

    const { container } = render(<PinInput />)

    const input = container.querySelector('input')

    expect(input?.attributes).toHaveProperty('disabled')
  })

  it("should show alert when there's an error", () => {
    usePinInputSpy.mockReturnValueOnce({
      pin: '',
      setPin: setPinSpy,
      loading: false,
      error: true,
    })

    const { getByText } = render(<PinInput />)

    expect(getByText('Pin is incorrect')).toBeDefined()
  })
})
