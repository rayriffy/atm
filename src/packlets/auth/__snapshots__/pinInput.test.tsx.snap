// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PinInput /> > should render and match snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-48cong"
  >
    <input
      aria-label="Please enter your pin code"
      autocomplete="off"
      class="chakra-pin-input css-0"
      data-index="0"
      id="pin-input-:r0:-0"
      inputmode="numeric"
      placeholder="○"
      type="password"
      value=""
    />
    <input
      aria-label="Please enter your pin code"
      autocomplete="off"
      class="chakra-pin-input css-0"
      data-index="1"
      id="pin-input-:r0:-1"
      inputmode="numeric"
      placeholder="○"
      type="password"
      value=""
    />
    <input
      aria-label="Please enter your pin code"
      autocomplete="off"
      class="chakra-pin-input css-0"
      data-index="2"
      id="pin-input-:r0:-2"
      inputmode="numeric"
      placeholder="○"
      type="password"
      value=""
    />
    <input
      aria-label="Please enter your pin code"
      autocomplete="off"
      class="chakra-pin-input css-0"
      data-index="3"
      id="pin-input-:r0:-3"
      inputmode="numeric"
      placeholder="○"
      type="password"
      value=""
    />
  </div>
</div>
`;
